From 6d74a3ac208705e4ec21b0f2436170d23d53e89e Mon Sep 17 00:00:00 2001
From: Robert Griebl <robert.griebl@qt.io>
Date: Sat, 10 Oct 2026 01:14:18 +0200
Subject: [PATCH] Wayland: fix crash when parsing wl_keyboard keymap

Commit 6a0823f53045c4d16cf592f0118951b84a5c4728 introduced a crash on
older libxkbcommon (e.g. Ubuntu 22.04 and RHEL 9.6), where a 0 byte
anywhere in the map string - even at the end - leads to an internal
library error and a nullptr return.

Also, we cannot dump hundreds of lines of keymap data with "warning"
priority to the normal "wayland" category.

Change-Id: Ib3464578ecd07e57ce074462ab2a649f86eac29c
Fixes: QTBUG-151282
Reviewed-by: David Redondo <qt@david-redondo.de>
(cherry picked from commit f73e18f1581568c78bd996b83c28df519610cbcd)
Origin: upstream, https://invent.kde.org/qt/qt/qtbase/-/commit/6d74a3ac208705e4ec21b0f2436170d23d53e89e
Forwarded: not-needed
Last-Update: 2026-10-29
---
 src/plugins/platforms/wayland/qwaylandinputdevice.cpp | 18 +++++++++++++-----
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/src/plugins/platforms/wayland/qwaylandinputdevice.cpp b/src/plugins/platforms/wayland/qwaylandinputdevice.cpp
index 960ebe8ab81..24621dba017 100644
--- a/src/plugins/platforms/wayland/qwaylandinputdevice.cpp
+++ b/src/plugins/platforms/wayland/qwaylandinputdevice.cpp
@@ -1254,10 +1254,22 @@ void QWaylandInputDevice::Keyboard::keyboard_keymap(uint32_t format, int32_t fd,
     if (map_str == MAP_FAILED)
         return;
 
-    mXkbKeymap.reset(xkb_keymap_new_from_string(mParent->mQDisplay->xkbContext(), map_str,
+    // We need to be able to cope with a non 0-terminated map_str, but we also need to make
+    // sure that map_str does not contain any 0-bytes at all, as older libxkbcommon cannot
+    // deal with those and fail (e.g. RHEL 9.6 and Ubuntu 22.04)
+    mXkbKeymap.reset(xkb_keymap_new_from_buffer(mParent->mQDisplay->xkbContext(),
+                                                map_str,
+                                                strnlen(map_str, size),
                                                 XKB_KEYMAP_FORMAT_TEXT_V1,
                                                 XKB_KEYMAP_COMPILE_NO_FLAGS));
-    QXkbCommon::verifyHasLatinLayout(mXkbKeymap.get());
+
+    if (mXkbKeymap) {
+        QXkbCommon::verifyHasLatinLayout(mXkbKeymap.get());
+    } else {
+        qCDebug(lcQpaWaylandInput) << "failed to compile a keymap, size =" << size
+                                   << ", keymap string:\n"
+                                   << QByteArrayView(map_str, strnlen(map_str, size));
+    }
 
     munmap(map_str, size);
 
